### PR TITLE
[Fiber] Make error handling more resilient

### PR DESCRIPTION
**Pull request by gaearon**
Supersedes #8166.

* Create a separate test file for incremental error handling 
* Ensure that errors in one root don't prevent work on another root
* Fix an issue where boundary state change would get ignored in incremental mode

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 1]
```diff
@@ -445,74 +446,100 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
-        nextUnitOfWork = findNextUnitOfWork();
+        if (!nextUnitOfWork) {
+          nextUnitOfWork = findNextUnitOfWork();
+        }
       }
+      if (nextUnitOfWork) {
+        scheduleCallbackAtPriority(nextPriorityLevel);
+      }
+    } finally {
+      isPerformingTaskWork = false;
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Functions that contain a try-catch/finally can't be optimized in V8. Can we move this reset back to the catch in `handleErrors`?

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 2]
```diff
@@ -561,47 +603,28 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
       try {
         performTaskWorkUnsafe();
       } catch (error) {
-        isPerformingTaskWork = false;
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: I moved this one into a try/finally in performTaskWorkUnsafe so that it's only set there.

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 3]
```diff
@@ -561,47 +603,28 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
         trapError(nextUnitOfWork, error, false);
       }
     }
 
+    // Unschedule the roots that fataled so that we can later schedule
+    // other updates to the same roots and potentially replace them.
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Can you explain in more detail why it's necessary to unschedule failed roots?

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 1 above)
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Usual strategy is to split the try-catch/finally block into a separate function that does nothing but an unsafe function. We should do a pass at some point to do this across the codebase.

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 2 above)
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: I agree it's easier to follow this way but I don't think it's worth a perf hit

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 1 above)
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: I know but we're completely moving this code around every week. I'm worried microoptimizations just make it hard to follow right now. Let's do a pass as soon as it stabilizes a little bit?

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 3 above)
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: I want to continue working on the other roots in the same frame even if some failed, and only throw errors after all work is done (or frame is over). If I don't unschedule failed roots, `findNextUnitOfWork` will find them again, and we'll get into an infinite loop handling their errors again, then finding them again, and so on.

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 1 above)
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Haha good point

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Maybe drop a TODO in there?

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 4]
```diff
@@ -611,15 +634,91 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
     }
     nextTrappedErrors.push({
-      boundary: fiber ? findClosestErrorBoundary(fiber) : null,
+      boundary,
       error,
+      root,
     });
   }
 
+  function unscheduleWork(unscheduledRoots : Set<FiberRoot>) {
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: How would I know which root to work on?

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 1 above)
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: 👍 Could you please?

Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: When we do it, we should do it consistently IMO, not in an ad hoc way. Use a `tryCatch` helper like Bluebird or something like this.

**Comment by acdlite** [comment]
Going to try to remove the need for unscheduling for a bit, otherwise I'll merge as-is

**Comment by gaearon** [comment]
👍 Leaving it to you.
